CONTRIBUTING.md

```diff
@@ -29,3 +29,125 @@ Recommendations and requirements for how to best contribute to Parquet. We striv
+   be [parquet-java](http://github.com/apache/parquet-java).  It is preferred
+   that the second implementation be
+   [parquet-cpp](https://github.com/apache/arrow) or
+   [parquet-rs](https://github.com/apache/arrow-rs), however at the discretion
+   of the PMC any open source Parquet implementation may be acceptable.
+   Implementations whose contributors actively participate in the community
+   (e.g. keep their feature matrix up-to-date on parquet-site) are more likely
+   to be considered. If discussed as a requirement in step one demonstration
+   of integration with a query engine is also required for this step.
+
+Unless otherwise discussed, it is expected the implementations will develop from
+the main branch (i.e. backporting is not expected).
+
+3. After the first two steps are complete a formal vote is held on the Parquet
+   mailing list to officially ratify the feature.  After the vote passes the
+   format change is merged into the parquet-format repository and it is expected
+   the changes from step 2 will also be merged soon after. Before merging into
+   Parquet-java a parquet-format release must be performed.
+
+#### General guidelines/preferences on additions.
+
+1. To the greatest extent possible changes should have an option for forwards
+   compatibility (old readers can still read files).
+2. New encodings should be fully specified in this repository and ideally not
+   rely on an external dependencies for implementation (i.e. Parquet is the
+   source of truth for the encoding).
+3. New compression mechanisms must have a pure Java implementation that can be
+   used as dependency in parquet-java.
+
+### Releases
+
+The Parquet community aims to do releases of the format package only as needed
+when new features are introduced. If multiple new features are being proposed
+simultaneously some features might be consolidated into the same release.
+Guidance is provided below on when implementations should enable features added
+to the specification.  Due to confusion in the past over parquet versioning it
+is not expected that there will be a 3.0 release of the specification in the
+foreseeable future.
+
+### Compatibility and Feature Enablement
+
+For the purposes of this discussion we classify features into the following buckets:
+
+1. Backwards compatible. A file written under an older version of the format
+   should be readable under a newer version of the format.
+2. Forwards compatible. A file written under a newer version of the format with
+   the enabled feature can be read under an older version of the format, but
+   some information might be missing or performance might be suboptimal.
+3. Forward incompatible. A file written under a new version of the format with
+   the feature enabled cannot be read under and older version of the format
+   (e.g. Adding a new compression algorithm.
+
+The Parquet community hopes that new features are widely beneficial to users of
```

Comment:
I think trying to create a process to roll out a feature across all implementations is an admirable goal, however I am not sure there is sufficient consensus to include it in this document. If we include this (aspirational) process I fear there is a non trivial chance it would not be followed in practice, and thus its inclusion would be confusing

Thus I suggest we consider not including the rollout in docs

Reply:
I tried to remove almost all prescriptions that actually require effort out of the suggestions.  I feel that if we don't provide some guidance the ecosystem will never really move forward.  If possible it would be nice to focus on concrete concerns.  Are there any parts of what is documented here that you see as problematic as a maintainer from arrow-rs?  Also @pitrou anything you see as specifically problematic from a parquet-cpp perspective?

Reply:
Thank you for the update. However, it is my opinion that the updated text still adds more potential for confusion than it adds. Specifically trying to proscribe the rollout cadence across loosly (at best) coordinated implementations in a document is unlikely to change behaviors and is likely to create confusion

> Are there any parts of what is documented here that you see as problematic as a maintainer from arrow-rs?

One specific example is the mention of feature flags. While I am sure we could over come the technical details, as a maintainer I would not want to be in the position to try and dictate when a featre should be used by our users -- instead I would imagine that as new features were supported in parquet-rs there would be an appropriate writer configuration that users would decide to enable/disable as they deemed appropriate. 

Another potential improvement (orthogonal to this disucssion) would be to label new format additions with the date in which they were added to the spec, and perhaps track the date when support was added to other implementations

Reply:
> Specifically trying to proscribe the rollout cadence across loosly (at best) coordinated implementations in a document is unlikely to change behaviors and is likely to create confusion

Is there specific language here that you find problematic here? With the latest revision, my intent was to make the recommendation only about preserving compatibility.  i.e. forward incompatible changes must wait at least 2 years before an implementation decides to turn them on by default.  I wonder if changing the phrasing from MAY to MUST NOT USE before would help clarify?  

> instead I would imagine that as new features were supported in parquet-rs there would be an appropriate writer configuration that users would decide to enable/disable as they deemed appropriate.

I agree.  Maybe this is a terminology issue.  I view feature flag as a type of configuration, and I think I removed all recommendations for when, if ever, an implementation should turn something on by default (please let me know if something is still ambiguous or maybe I missed something in rephrasing).


> Another potential improvement (orthogonal to this disucssion) would be to label new format additions with the date in which they were added to the spec, and perhaps track the date when support was added to other implementations

Agreed, I'll add this to the process I thought the feature matrix that was already underreview would be a good place for this?  I would like to avoid putting too many dates directly in the spec to avoid confusion.

Reply:
I'll try to rephrase to make the issues above clearer.

> I also think the major version bump suggestion is so library specific it is not worth putting in the guidance. The libraries have their own cadence (e.g. parquet-rs and parquet-cpp have major releases based on calendar cycles rather than feature content)

My intent here was not to specify a release cadence but to specify that if changing a forward incompatible change to default the next release must be a major version release, I will try to clarify.  I think parquet-rs and parquet-cpp trivially satisfy this given the current release policies.

Reply:
@alamb thanks for feedback, I've update to clarify and remove upper case for SHOULD/MAY language, please let me know if the current iteration still raises concerns.